### PR TITLE
preconditioned conjugate gradient and conjugate residual

### PR DESCRIPTION
**Pull request by Xx-fractale-xX**
issue #52

### CODE REVIEW

File: src/cg.jl [hunk 1]
```diff
@@ -64,7 +69,8 @@ function cg{T <: Number}(A :: AbstractLinearOperator, b :: AbstractVector{T};
 
     @kaxpy!(n,  α,  p, x)
     @kaxpy!(n, -α, Ap, r)
-    γ_next = @kdot(n, r, r)
+    z = M * z
```
Comment (dpo) [inline src/cg.jl]: `z = M * r` ?

**Comment by dpo** [comment]
I have to check preconditioned CR. We'll also need some tests.

**Comment by dpo** [comment]
Ps: please use `fixes #52` in your commit message.

**Comment by Xx-fractale-xX** [comment]
I use ```fixes #52``` now, but I want to add preconditioners in other algorithms before. It's why I didn't want to close the issue.
For tests, I was thinking about Jacobi preconditioner,  circulant matrix for symmetric Toeplitz matrix and LLDL factorization from ```JuliaSmoothOptimizers/LLDL.jl```.

**Comment by dpo** [comment]
> I use fixes #52 now, but I want to add preconditioners in other algorithms before. It's why I didn't want to close the issue.

Ok, but there's no need to fix all methods in one PR. It's easier to review small targeted PRs than large ones. I recommend updating #52 with the list of all methods and tick off those that already accept a preconditioner. That way, we'll track progress.

> For tests, I was thinking about Jacobi preconditioner, circulant matrix for symmetric Toeplitz matrix and LLDL factorization from JuliaSmoothOptimizers/LLDL.jl.

It's ok to keep it simple in tests. A positive definite diagonal preconditioner will do, but feel free to add others. A test with LLDL preconditioner might fit better in LLDL.jl than here?!

**Comment by Xx-fractale-xX** [comment]
We updated `tired` before `iter`. If we didn't converged before itmax, we stayed in while boucle one more time. It can distort some benchmarks...

**Comment by abelsiqueira** [comment]
Thank you

**Comment by dpo** [comment]
Hold on, I haven't even had a chance to check this thoroughly.

**Comment by abelsiqueira** [comment]
Ok, I'm reverting this.

**Comment by abelsiqueira** [comment]
Reverted by hand, and opened the PR from the same @Xx-fractale-xX branch

**Pull request merged**